### PR TITLE
Add  `orientation` parameter to Boxplot and deprecate `vert`

### LINKED ISSUE #13435: boxplot/violinplot orientation-setting API
### Bug report

**Bug summary**

Currently, boxplot(), bxp(), violin(), and violinplot() take a `vert: bool` argument to switch between vertical and horizontal; this is inconsistent with colorbar(), hist(), eventplot(), and Slider(), which take an `orientation: {"vertical", "horizontal"}` argument.

For consistency, I'd suggest adding support for `orientation` (both because I think colorbar() and hist(), in particular, are probably the most used among all these, and because I like the API better :-)), and later deprecate `vert`.  Thoughts?

attn @phobson who has been involved in boxplots, IIRC.

### PR DESCRIPTION
**Pull request by saranti**
## PR summary

Closes #13435 together with PR #27998.
Replace the vert: bool parameter with orientation : {'vertical', 'horizontal'}. Changes are very similar to the work done on #27998 and most of it works interchangeably. 

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [x] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by saranti** [comment]
That test is passing on my end, I don't know what the problem is 🤔

**Comment by rcomer** [comment]
Looks like the test just timed out.  Let’s see if a re-spin helps…

File: lib/matplotlib/axes/_axes.py [hunk 1]
```diff
@@ -4047,8 +4060,6 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
                                        labels=tick_labels, autorange=autorange)
         if notch is None:
             notch = mpl.rcParams['boxplot.notch']
-        if vert is None:
-            vert = mpl.rcParams['boxplot.vertical']
```
Comment (timhoffm) [inline lib/matplotlib/axes/_axes.py:4051]: We need a deprecation on `mpl.rcParams['boxplot.vertical']` as well. This is a bit tricky.

1) We don't have a direct mechanism for deprecating `rcParams`. Likely a custom validator similar to https://github.com/matplotlib/matplotlib/blob/8fb842c78f1b57b13a3d0fe63e83f3d31694f887/lib/matplotlib/rcsetup.py#L441 is the way to go.
2) Determining priority: Not completely thought through, but I believe, if `mpl.rcParams['boxplot.vertical'] = False` that should take precedence over `orientation`. `mpl.rcParams['boxplot.vertical'] = True` should likely be ignored. That should give expected results for reasonable use cases. - We cannot easily detect whether the rcParam and `orientation` were explicitly set or use their defaults. That detection would require extra machinery (i.e. an extra None=not set state), which we probably can get away without folling the above rule.

Comment (QuLogic) [reply lib/matplotlib/axes/_axes.py:4051]: We do have some rcParams deprecation mechanism here:
https://github.com/matplotlib/matplotlib/blob/ec0653d9f54ca5c2ae6d5e8f84daaed8fe178886/lib/matplotlib/__init__.py#L637-L646

**Comment by saranti** [comment]
- Now `vert=True` won't emit a deprecation warning, regardless of how it's set. It also won't override `orientation`. 

- `vert=False` will emit a deprecation warning and override `orientation`.

- Setting `mpl.rcParams['boxplot.vertical']` to False will trigger an rcparam deprecation warning.

File: doc/api/next_api_changes/deprecations/28074-TS.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+``boxplot`` and ``bxp`` *vert* parameter
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The parameter *vert: bool* and the *boxplot.vertical* rcparam has
+been deprecated on `~.Axes.boxplot` and `~.Axes.bxp`.
+It is replaced by *orientation: {"vertical", "horizontal"}* for API
+consistency.
```
Comment (timhoffm) [inline doc/api/next_api_changes/deprecations/28074-TS.rst:7]: Please add the deprecatio of the rcParam here as well.

File: lib/matplotlib/__init__.py [hunk 3]
```diff
@@ -732,6 +732,11 @@ def __setitem__(self, key, val):
                 if val is rcsetup._auto_backend_sentinel:
                     if 'backend' in self:
                         return
+            elif key == 'boxplot.vertical' and not val:
```
Comment (timhoffm) [inline lib/matplotlib/__init__.py:735]: I'm not clear how the internals of of `rcParams` work together but, in a test this does not raise a warning when the parameter is set in matplotlibrc or when set via `plt.rc_context`. Maybe it's simpler to put the warning to where the value is read in `bxp`?

Either way, I don't care too much about how this is implemented, but I would like to see a test that ensures the warning on the rcParam is issued.

File: doc/api/next_api_changes/deprecations/28074-TS.rst (hunk 2 above)
Comment (saranti) [reply doc/api/next_api_changes/deprecations/28074-TS.rst:7]: Do you mean in the heading? or under its own heading?

File: lib/matplotlib/__init__.py (hunk 3 above)
Comment (saranti) [reply lib/matplotlib/__init__.py]: The warnings are suppressed for some reason when it's set that way. I just put it in `bxp`

File: doc/api/next_api_changes/deprecations/28074-TS.rst (hunk 2 above)
Comment (timhoffm) [reply doc/api/next_api_changes/deprecations/28074-TS.rst:7]: You can do it under this heading.

File: doc/api/next_api_changes/deprecations/28074-TS.rst [hunk 4]
```diff
@@ -0,0 +1,9 @@
+``boxplot`` and ``bxp`` *vert* parameter, and ``boxplot.vertical`` rcparam
```
Comment (timhoffm) [inline doc/api/next_api_changes/deprecations/28074-TS.rst:1]: Using the actual pattern is easier to recognize:
```suggestion
``boxplot`` and ``bxp`` *vert* parameter, and ``rcParams["boxplot.vertical"]``
```

File: doc/api/next_api_changes/deprecations/28074-TS.rst [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The parameter *vert: bool* has been deprecated on `~.Axes.boxplot` and
+`~.Axes.bxp`. It is replaced by *orientation: {"vertical", "horizontal"}*
+for API consistency.
+
+The *boxplot.vertical* rcparam which controlled the orientation of ``boxplot``
+is deprecated as the new API does not require its use.
```
Comment (timhoffm) [inline doc/api/next_api_changes/deprecations/28074-TS.rst:9]:
```suggestion
``rcParams["boxplot.vertical"]``, which controlled the orientation of ``boxplot``,
is deprecated without replacement.
```
It's an idependent design decision that it's not reasonable/necessary to control the orientation via rcParams.

**Review by timhoffm — APPROVED** [review]
Modulo two documentation suggestions.

Thanks @saranti!

File: lib/matplotlib/tests/test_axes.py [hunk 6]
```diff
@@ -3155,7 +3155,7 @@ def _bxp_test_helper(
     logstats = mpl.cbook.boxplot_stats(
         np.random.lognormal(mean=1.25, sigma=1., size=(37, 4)), **stats_kwargs)
     fig, ax = plt.subplots()
-    if bxp_kwargs.get('vert', True):
+    if not bxp_kwargs.get('orientation') or bxp_kwargs.get('orientation') == 'vertical':
```
Comment (rcomer) [inline lib/matplotlib/tests/test_axes.py:3158]:
```suggestion
    if bxp_kwargs.get('orientation', 'vertical') == 'vertical':
```
Defines a default and so avoids calling `get` twice.

**Review by rcomer — COMMENTED** [review]
Thanks @saranti, this is looking great!

It took me a while to get my head around why we are not warning when `vert=True` is explicitly passed.  I guess it's because it's too complicated to do for `bxp` as we can't tell whether it came from the rcParam.  However would it be worth adding an extra warning in `boxplot`specifically for when `vert=True`?  Apologies if I've missed some discussion on this somewhere.

**Comment by timhoffm** [comment]
> Thanks @saranti, this is looking great!
> 
> It took me a while to get my head around why we are not warning when `vert=True` is explicitly passed. I guess it's because it's too complicated to do for `bxp` as we can't tell whether it came from the rcParam. However would it be worth adding an extra warning in `boxplot`specifically for when `vert=True`? Apologies if I've missed some discussion on this somewhere.

Good catch. Since we have the tri-state True/False/None, we can detect whenever True/False was passed explicitly. If we push rcParams resolution down to `bxp()` we can even detect that also for `bxp()`.

**Comment by rcomer** [comment]
> If we push rcParams resolution down to bxp() we can even detect that also for bxp().

That would mean the rcParam affects cases where the user calls `bxp` directly, which it didn’t before.  However I think it’s a little odd that it didn’t, so we could consider that a bugfix 🤔

**Comment by rcomer** [comment]
Could we also set the rcParam in the default matplotlibrc to `None` and then add it to this dictionary (highlighted by @QuLogic above)?
https://github.com/matplotlib/matplotlib/blob/46b39ab5bea6668e56140da568bde60276f6f906/lib/matplotlib/__init__.py#L646

If I have understood, that would mean we get the warning if the rcParam is explicitly set to either `True` or `False` by the user.

**Comment by timhoffm** [comment]
> > If we push rcParams resolution down to bxp() we can even detect that also for bxp().
> 
> That would mean the rcParam affects cases where the user calls `bxp` directly, which it didn’t before. However I think it’s a little odd that it didn’t, so we could consider that a bugfix 🤔

Indeed. But it would only affect `bxp()` calls, where `vert` is not passed *and* the rcParam is modified. I'd be willing to accept that change because being able to warn on all explicit `bxp(..., vert=)` is IMHO more important. And as you say, it can be considered a bugfix. And the  rcParam gets deprecated, so the user has to adapt their code anyway.

**Comment by saranti** [comment]
I didn't see any reason why the `vert = mpl.rcParams['boxplot.vertical']` needed to be in `boxplot` so I moved it to `bxp`. Now it also gives a warning to `bxp` calls where it previously didn't.
Explicitly setting `vert=True` will now also emit a deprecation warning.

**Review by timhoffm — APPROVED** [review]

**Review by rcomer — APPROVED** [review]

**Pull request merged**